lib/libsdp/sdp.3
.\" Copyright (c) 2003 Maksim Yevmenkin <m_evmenkin@yahoo.com>
.\" All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED. IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.\" $Id: sdp.3,v 1.1 2003/09/07 20:34:19 max Exp $
.\" $FreeBSD: stable/11/lib/libsdp/sdp.3 267773 2014-06-23 08:23:05Z bapt $
.\"
.Dd May 27, 2005
.Dt SDP 3
.Os
.Sh NAME
.Nm SDP_GET8 ,
.Nm SDP_GET16 ,
.Nm SDP_GET32 ,
.Nm SDP_GET64 ,
.Nm SDP_GET128 ,
.Nm SDP_GET_UUID128 ,
.Nm SDP_PUT8 ,
.Nm SDP_PUT16 ,
.Nm SDP_PUT32 ,
.Nm SDP_PUT64 ,
.Nm SDP_PUT128 ,
.Nm SDP_PUT_UUID128 ,
.Nm sdp_open ,
.Nm sdp_open_local ,
.Nm sdp_close ,
.Nm sdp_error ,
.Nm sdp_search ,
.Nm sdp_attr2desc ,
.Nm sdp_uuid2desc
.Nd Bluetooth SDP routines
.Sh LIBRARY
.Lb libsdp
.Sh SYNOPSIS
.In bluetooth.h
.In sdp.h
.Fn SDP_GET8 "b" "cp"
.Fn SDP_GET16 "s" "cp"
.Fn SDP_GET32 "l" "cp"
.Fn SDP_GET64 "l" "cp"
.Fn SDP_GET128 "l" "cp"
.Fn SDP_GET_UUID128 "l" "cp"
.Fn SDP_PUT8 "b" "cp"
.Fn SDP_PUT16 "s" "cp"
.Fn SDP_PUT32 "l" "cp"
.Fn SDP_PUT64 "l" "cp"
.Fn SDP_PUT128 "l" "cp"
.Fn SDP_PUT_UUID128 "l" "cp"
.Ft "void *"
.Fn sdp_open "bdaddr_t const *l" "bdaddr_t const *r"
.Ft "void *"
.Fn sdp_open_local "char const *control"
.Ft int32_t
.Fn sdp_close "void *xs"
.Ft int32_t
.Fn sdp_error "void *xs"
.Ft int32_t
.Fo sdp_search
.Fa "void *xs" "uint32_t plen" "uint16_t const *pp" "uint32_t alen"
.Fa "uint32_t const *ap" "uint32_t vlen" "sdp_attr_t *vp"
.Fc
.Ft "char const * const"
.Fn sdp_attr2desc "uint16_t attr"
.Ft "char const * const"
.Fn sdp_uuid2desc "uint16_t uuid"
.Ft int32_t
.Fo sdp_register_service
.Fa "void *xss" "uint16_t uuid" "bdaddr_p const bdaddr" "uint8_t const *data"
.Fa "uint32_t datalen" "uint32_t *handle"
.Fc
.Ft int32_t
.Fn sdp_unregister_service "void *xss" "uint32_t handle"
.Ft int32_t
.Fo sdp_change_service
.Fa "void *xss" "uint32_t handle" "uint8_t const *data" "uint32_t datalen"
.Fc
.Sh DESCRIPTION
The
.Fn SDP_GET8 ,
.Fn SDP_GET16 ,
.Fn SDP_GET32 ,
.Fn SDP_GET64
and
.Fn SDP_GET128
macros are used to get byte, short, long, long long and 128-bit integer
from the buffer pointed by
.Fa cp
pointer.
The pointer is automatically advanced.
.Pp
The
.Fn SDP_PUT8 ,
.Fn SDP_PUT16 ,
.Fn SDP_PUT32 ,
.Fn SDP_PUT64
and
.Fn SDP_PUT128
macros are used to put byte, short, long, long long and 128-bit integer
into the buffer pointed by
.Fa cp
pointer.
The pointer is automatically advanced.
.Pp
.Fn SDP_GET_UUID128
and
.Fn SDP_PUT_UUID128
macros are used to get and put 128-bit UUID into the buffer pointed by
.Fa cp
pointer.
The pointer is automatically advanced.
.Pp
The
.Fn sdp_open
and
.Fn sdp_open_local
functions each return a pointer to an opaque object describing SDP session.
The
.Fa l
argument passed to
.Fn sdp_open
function should point to a source BD_ADDR.
If source BD_ADDR is
.Dv NULL
then source address
.Dv NG_HCI_BDADDR_ANY
is used.
The
.Fa r
argument passed to
.Fn sdp_open
function should point to a
.Pf non- Dv NULL
remote BD_ADDR.
Remote BD_ADDR cannot be
.Dv NG_HCI_BDADDR_ANY .
The
.Fn sdp_open_local
function takes path to the control socket and opens a connection to a local
SDP server.
If path to the control socket is
.Dv NULL
then default
.Pa /var/run/sdp
path will be used.
.Pp
The
.Fn sdp_close
function terminates active SDP session and deletes SDP session object.
The
.Fa xs
parameter should point to a valid SDP session object created with
.Fn sdp_open
or
.Fn sdp_open_local .
.Pp
The
.Fn sdp_error
function returns last error that is stored inside SDP session object.
The
.Fa xs
parameter should point to a valid SDP session object created with
.Fn sdp_open
or
.Fn sdp_open_local .
The error value returned can be converted to a human readable message by
calling
.Xr strerror 3
function.
.Pp
The
.Fn sdp_search
function is used to perform SDP Service Search Attribute Request.
The
.Fa xs
parameter should point to a valid SDP session object created with
.Fn sdp_open
or
.Fn sdp_open_local .
The
.Fa pp
parameter is a Service Search Pattern - an array of one or more Service
Class IDs.
The maximum number of Service Class IDs in the array is 12.
The
.Fa plen
parameter is the length of the Service Search pattern.
The
.Fa ap
parameter is an Attribute ID Range List - an array of one or more SDP Attribute
ID Range.
Each attribute ID Range is encoded as a 32-bit unsigned integer data
element, where the high order 16 bits are interpreted as the beginning
attribute ID of the range and the low order 16 bits are interpreted as the
ending attribute ID of the range.
The attribute IDs contained in the Attribute ID Ranges List must be listed in
ascending order without duplication of any attribute ID values.
Note that all attributes may be requested by specifying a range of
0x0000-0xFFFF.
The
.Fa alen
parameter is the length of the Attribute ID Ranges List.
The
.Fn SDP_ATTR_RANGE "lo" "hi"
macro can be used to prepare Attribute ID Range.
The
.Fa vp
parameter should be an array of
.Vt sdp_attr_t
structures.
Each
.Vt sdp_attr_t
structure describes single SDP attribute and defined as follows:
.Bd -literal -offset indent
struct sdp_attr {
        uint16_t        flags;
#define SDP_ATTR_OK             (0 << 0)
#define SDP_ATTR_INVALID        (1 << 0)
#define SDP_ATTR_TRUNCATED      (1 << 1)
        uint16_t        attr;  /* SDP_ATTR_xxx */
        uint32_t        vlen;  /* length of the value[] in bytes */
        uint8_t        *value; /* base pointer */
};
typedef struct sdp_attr         sdp_attr_t;
typedef struct sdp_attr *       sdp_attr_p;
.Ed
.Pp
The caller of the
.Fn sdp_search
function is expected to prepare the array of
.Vt sdp_attr
structures and for each element of the array both
.Va vlen
and
.Va value
must be set.
The
.Fn sdp_search
function will fill each
.Vt sdp_attr_t
structure with attribute and value, i.e., it will set
.Va flags ,
.Va attr
and
.Va vlen
fields.
The actual value of the attribute will be copied into
.Va value
buffer.
Note: attributes are returned in the order they appear in the Service Search
Attribute Response.
SDP server could return several attributes for the same record.
In this case the order of the attributes will be: all attributes for the first
records, then all attributes for the second record etc.
.Pp
The
.Fn sdp_attr2desc
and
.Fn sdp_uuid2desc
functions each take a numeric attribute ID/UUID value and convert it to a
human readable description.
.Pp
The
.Fn sdp_register_service
function
is used to register service with the local SDP server.
The
.Fa xss
parameter should point to a valid SDP session object obtained from
.Fn sdp_open_local .
The
.Fa uuid
parameter is a SDP Service Class ID for the service to be registered.
The
.Fa bdaddr
parameter should point to a valid BD_ADDR.
The service will be only advertised if request was received by the local device
with
.Fa bdaddr .
If
.Fa bdaddr
is set to
.Dv NG_HCI_BDADDR_ANY
then the service will be advertised to any remote devices that queries for it.
The
.Fa data
and
.Fa datalen
parameters specify data and size of the data for the service.
Upon successful return
.Fn sdp_register_service
will populate
.Fa handle
with the SDP record handle.
This parameter is optional and can be set to
.Dv NULL .
.Pp
The
.Fn sdp_unregister_service
function
is used to unregister service with the local SDP server.
The
.Fa xss
parameter should point to a valid SDP session object obtained from
.Fn sdp_open_local .
The
.Fa handle
parameter should contain a valid SDP record handle of the service to be
unregistered.
.Pp
The
.Fn sdp_change_service
function is used to change data associated with the existing service on
the local SDP server.
The
.Fa xss
parameter should point to a valid SDP session object obtained from
.Fn sdp_open_local .
The
.Fa handle
parameter should contain a valid SDP record handle of the service to be changed.
The
.Fa data
and
.Fa datalen
parameters specify data and size of the data for the service.
.Sh CAVEAT
When registering services with the local SDP server the application must
keep the SDP session open.
If SDP session is closed then the local SDP server will remove all services
that were registered over the session.
The application is allowed to change or unregister service if it was registered
over the same session.
.Sh EXAMPLES
The following example shows how to get
.Dv SDP_ATTR_PROTOCOL_DESCRIPTOR_LIST
attribute for the
.Dv SDP_SERVICE_CLASS_SERIAL_PORT
service from the remote device.
.Bd -literal -offset indent
bdaddr_t       remote;
uint8_t        buffer[1024];
void          *ss    = NULL;
uint16_t       serv  = SDP_SERVICE_CLASS_SERIAL_PORT;
uint32_t       attr  = SDP_ATTR_RANGE(
                            SDP_ATTR_PROTOCOL_DESCRIPTOR_LIST,
                            SDP_ATTR_PROTOCOL_DESCRIPTOR_LIST);
sdp_attr_t     proto = { SDP_ATTR_INVALID,0,sizeof(buffer),buffer };

/* Obtain/set remote BDADDR here */

if ((ss = sdp_open(NG_HCI_BDADDR_ANY, remote)) == NULL)
        /* exit ENOMEM */
if (sdp_error(ss) != 0)
        /* exit sdp_error(ss) */

if (sdp_search(ss, 1, &serv, 1, &attr, 1, &proto) != 0)
        /* exit sdp_error(ss) */

if (proto.flags != SDP_ATTR_OK)
        /* exit see proto.flags for details */

/* If we got here then we have attribute value in proto.value */
.Ed
.Sh DIAGNOSTICS
Both
.Fn sdp_open
and
.Fn sdp_open_local
will return
.Dv NULL
if memory allocation for the new SDP session object fails.
If the new SDP object was created then caller is still expected to call
.Fn sdp_error
to check if there was connection error.
.Pp
The
.Fn sdp_search ,
.Fn sdp_register_service ,
.Fn sdp_unregister_service
and
.Fn sdp_change_service
functions return non-zero value on error.
The caller is expected to call
.Fn sdp_error
to find out more about error.
.Sh SEE ALSO
.Xr bluetooth 3 ,
.Xr strerror 3 ,
.Xr sdpcontrol 8 ,
.Xr sdpd 8
.Sh AUTHORS
.An Maksim Yevmenkin Aq Mt m_evmenkin@yahoo.com
.Sh BUGS
Most likely.
Please report bugs if found.


NEWS
GNU diffutils NEWS                                    -*- outline -*-

* Noteworthy changes in release 3.2 (2011-09-02) [stable]

** Changes in behavior

  --ignore-file-name-case now applies at the top level too.
  For example, "diff dir inIt" might compare "dir/Init" to "inIt".

** New features

  diff and sdiff have a new option --ignore-trailing-space (-Z).

** Packaging

  The texinfo documentation no longer specifies "front-cover" or "back-cover"
  texts, so that it may now be included in Debian's "main" section.


* Noteworthy changes in release 3.1 (2011-08-10) [stable]

** Bug fixes

  diff no longer reports spurious differences merely because two entries
  in the same directory have names that compare equal in the current
  locale, or compare equal because --ignore-file-name-case was given.

* Noteworthy changes in release 3.0 (2010-05-03) [stable]

** Bug fixes

  diff once again prints the required "\ No newline at end of file" line
  when at least one input lacks a newline-at-EOF and the final hunk plus
  context-length aligns exactly with the end of the newline-lacking file.
  [bug introduced between 2.8.7 and 2.9]

** Changes in behavior

  In context-style diffs, diff prints a portion of a preceding "function"
  line for each hunk, with --show-function-line=RE (-F) or
  --show-c-function (-p).  Now, it trims leading blanks from such lines
  before extracting a prefix.  This is useful especially when a function
  line is so far indented that the name itself would be truncated or not
  included in the limited-width substring that diff appends.

  diff once again reports a difference with the diagnostic
  "Binary files A and B differ" when at least one of the files
  appears to be binary.  From 2.8.4 through diffutils-2.9, it printed
  "Files A and B differ".


* Noteworthy changes in release 2.9 (2010-02-11) [stable]

** New features

  New diff option --suppress-blank-empty.

  Bring back support for `diff -NUM', where NUM is a number,
  even when conforming to POSIX 1003.1-2001.  This change reverts to
  the behavior of GNU diff 2.7 and earlier.  This is a change only
  when conforming to POSIX 1003.1-2001; there is no effect when
  conforming to older POSIX versions.

  This change is in response to decisions taken in the January 2005
  Austin Group standardization meeting.  For more details, please see
  "Utility Syntax Guidelines" in the Minutes of the January 2005
  Meeting <http://www.opengroup.org/austin/docs/austin_239.html>.

  sdiff now understands '1' and '2' as synonyms for 'l' and 'r'.

** Changes in behavior

  sdiff and diff3 now invoke diff, not $(bindir)/diff

** Administrivia

  New discussion and bug-reporting address: bug-diffutils@gnu.org

  updated gnulib support


* Noteworthy changes in release 2.8.7 (2004-04-13) [stable]

  Version 2.8.7 contains no user-visible changes.

User-visible changes in version 2.8.6:

* New diff3 option --strip-trailing-cr.

* With -N and -P, inaccessible empty regular files (the kind of files
  that 'patch' creates to indicate nonexistent backups) are now
  treated as nonexistent when they are in the 'backup' file position.

* If multiple SKIP values are given to cmp, e.g., `cmp -i 10 -i 20',
  cmp now uses the maximal value instead of the last one.

* diff now omits the ".000000000" on hosts that do not support
  fractional time stamps.

Version 2.8.5 was not publicly released.

User-visible changes in version 2.8.4:

* Diff now simply prints "Files A and B differ" instead of "Binary
  files A and B differ".  The message is output if either A or B
  appears to be a binary file, and the old wording was misleading
  because it implied that both files are binary, which is not
  necessarily the case.

User-visible changes in version 2.8.3:

* New locale: en_US.

User-visible changes in version 2.8.2:

* New diff and sdiff option:
  --tabsize=COLUMNS
* If --ignore-space-change or --ignore-all-space is also specified,
  --ignore-blank-lines now considers lines to be empty if they contain
  only white space.
* More platforms now handle multibyte characters correctly when
  excluding files by name (diff -x and -X).
* New locales: hu, pt_BR.

User-visible changes in version 2.8.1:

* Documentation fixes.

User-visible changes in version 2.8:

* cmp and diff now conform to POSIX 1003.1-2001 (IEEE Std 1003.1-2001)
  if the underlying system conforms to POSIX and if the _POSIX2_VERSION
  environment variable is set to 200112.  Conformance removes support
  for `diff -NUM', where NUM is a number.  Use -C NUM or -U NUM instead.
* cmp now supports trailing operands SKIP1 and SKIP2, like BSD cmp.
* cmp -i or --ignore-initial now accepts SKIP1:SKIP2 option value.
* New cmp option: -n or --bytes.
* cmp's old -c or --print-chars option has been renamed;
  use -b or --print-bytes instead.
* cmp now outputs "byte" rather than "char" outside the POSIX locale.
* cmp -l's index column width now adjusts to fit larger (or smaller) files.
* cmp -l -s and cmp -s -l are not allowed.  Use cmp -s or cmp -l instead.
* diff uses ISO 8601 style time stamps for output times (e.g. "2001-11-23
  16:44:36.875702460 -0800") unless in the C or POSIX locale and the
  -c style is specified.
* diff's -I and -F options use the regexp syntax of grep, not of Emacs.
* diff now accepts multiple context arguments, and uses their maximum value.
* New diff and sdiff options:
  -E  --ignore-tab-expansion
  --strip-trailing-cr
* New diff options:
  --from-file=FILE, --to-file=FILE
  --ignore-file-name-case
  --no-ignore-file-name-case
* New diff3 and sdiff option:
  --diff-program=PROGRAM
* The following diff options are still accepted, but are no longer documented.
  They may be withdrawn in future releases.
  -h (omit; it has no effect)
  -H (use --speed-large-files instead)
  -L (use --label instead)
  -P (use --unidirectional-new-file instead)
  --inhibit-hunk-merge (omit; it has no effect)
* Recursive diffs now sort file names according to the LC_COLLATE locale

usr.sbin/bluetooth/btpand/sdp.c
/* $MidnightBSD$ */
/*	$NetBSD: sdp.c,v 1.2 2008/12/06 20:01:14 plunky Exp $	*/

/*-
 * Copyright (c) 2008 Iain Hibbert
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY EXPRESS OR
 * IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
 * OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
 * IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT, INDIRECT,
 * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT
 * NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
 * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
 * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
 * THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

/* $FreeBSD: stable/10/usr.sbin/bluetooth/btpand/sdp.c 187938 2009-01-30 22:23:21Z emax $ */

#include <sys/cdefs.h>
__RCSID("$NetBSD: sdp.c,v 1.2 2008/12/06 20:01:14 plunky Exp $");

#include <string.h>

#include "sdp.h"

/*
 * SDP data stream manipulation routines
 */

/* Bluetooth Base UUID */
static const uuid_t BASE_UUID = {
	0x00000000,
	0x0000,
	0x1000,
	0x80,
	0x00,
	{ 0x00, 0x80, 0x5f, 0x9b, 0x34, 0xfb }
};

/*
 * _sdp_match_uuid16(ptr, limit, uuid)
 *
 *	examine SDP data stream at ptr for a UUID, and return
 *	true if it matches the supplied short alias bluetooth UUID.
 *	limit is the first address past the end of valid data.
 */
bool
_sdp_match_uuid16(uint8_t **ptr, uint8_t *limit, uint16_t uuid)
{
	uint8_t *p = *ptr;
	uuid_t u1, u2;

	memcpy(&u1, &BASE_UUID, sizeof(uuid_t));
	u1.time_low = uuid;

	if (!_sdp_get_uuid(&p, limit, &u2)
	    || !uuid_equal(&u1, &u2, NULL))
		return false;

	*ptr = p;
	return true;
}

/*
 * _sdp_get_uuid(ptr, limit, uuid)
 *
 *	examine SDP data stream at ptr for a UUID, and extract
 *	to given storage, advancing ptr.
 *	limit is the first address past the end of valid data.
 */
bool
_sdp_get_uuid(uint8_t **ptr, uint8_t *limit, uuid_t *uuid)
{
	uint8_t *p = *ptr;

	if (p + 1 > limit)
		return false;

	switch (*p++) {
	case SDP_DATA_UUID16:
		if (p + 2 > limit)
			return false;

		memcpy(uuid, &BASE_UUID, sizeof(uuid_t));
		uuid->time_low = be16dec(p);
		p += 2;
		break;

	case SDP_DATA_UUID32:
		if (p + 4 > limit)
			return false;

		memcpy(uuid, &BASE_UUID, sizeof(uuid_t));
		uuid->time_low = be32dec(p);
		p += 4;
		break;

	case SDP_DATA_UUID128:
		if (p + 16 > limit)
			return false;

		uuid_dec_be(p, uuid);
		p += 16;
		break;

	default:
		return false;
	}

	*ptr = p;
	return true;
}

/*
 * _sdp_get_seq(ptr, limit, seq)
 *
 *	examine SDP data stream at ptr for a sequence. return
 *	seq pointer if found and advance ptr to next object.
 *	limit is the first address past the end of valid data.
 */
bool
_sdp_get_seq(uint8_t **ptr, uint8_t *limit, uint8_t **seq)
{
	uint8_t *p = *ptr;
	int32_t l;

	if (p + 1 > limit)
		return false;

	switch (*p++) {
	case SDP_DATA_SEQ8:
		if (p + 1 > limit)
			return false;

		l = *p;
		p += 1;
		break;

	case SDP_DATA_SEQ16:
		if (p + 2 > limit)
			return false;

		l = be16dec(p);
		p += 2;
		break;

	case SDP_DATA_SEQ32:
		if (p + 4 > limit)
			return false;

		l = be32dec(p);
		p += 4;
		break;

	default:
		return false;
	}
	if (p + l > limit)
		return false;

	*seq = p;
	*ptr = p + l;
	return true;
}

/*
 * _sdp_get_uint16(ptr, limit, value)
 *
 *	examine SDP data stream at ptr for a uint16_t, and
 *	extract to given storage, advancing ptr.
 *	limit is the first address past the end of valid data.
 */
bool
_sdp_get_uint16(uint8_t **ptr, uint8_t *limit, uint16_t *value)
{
	uint8_t *p = *ptr;
	uint16_t v;

	if (p + 1 > limit)
		return false;

	switch (*p++) {
	case SDP_DATA_UINT16:
		if (p + 2 > limit)
			return false;

		v = be16dec(p);
		p += 2;
		break;

	default:
		return false;
	}

	*value = v;
	*ptr = p;
	return true;
}
